packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -0,0 +1,542 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:async/async.dart';
+import 'package:convert/convert.dart';
+import 'package:crypto/crypto.dart';
+import 'package:meta/meta.dart';
+import 'package:pool/pool.dart';
+
+import '../base/file_system.dart';
+import '../base/platform.dart';
+import '../cache.dart';
+import '../convert.dart';
+import '../globals.dart';
+import 'exceptions.dart';
+import 'file_cache.dart';
+import 'source.dart';
+
+export 'source.dart';
+
+/// The function signature of a build target which can be invoked to perform
+/// the underlying task.
+typedef BuildInvocation = FutureOr<void> Function(
+    Map<String, ChangeType> inputs, Environment environment);
+
+/// A description of the update to each input file.
+enum ChangeType {
+  /// The file was added.
+  Added,
+  /// The file was deleted.
+  Removed,
+  /// The file was modified.
+  Modified,
+}
+
+/// Configuration for the build system itself.
+class BuildSystemConfig {
+  const BuildSystemConfig({this.resourcePoolSize});
+
+  /// The maximum number of concurrent tasks the build system will run.
+  ///
+  /// If not provided, defaults to [platform.numberOfProcessors].
+  final int resourcePoolSize;
+}
+
+/// A Target describes a single step during a flutter build.
+///
+/// The target inputs are required to be files discoverable via a combination
+/// of at least one of the environment values and zero or more local values.
+///
+/// To determine if a target needs to be executed, the [BuildSystem] performs
+/// a  hash of the file contents. This is tracked separately in the
+/// [FileCache].
+///
+/// The name of each stamp is the target name, joined with target platform and
```

Comment:
This is the first mention of 'stamp'. Maybe start with something like, 'For each target, executing its action creates a stamp file...'.